sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/FluxAutoComplete.java

```diff
@@ -0,0 +1,130 @@
+
+package com.azure.messaging.servicebus;
+
+import com.azure.core.util.logging.ClientLogger;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.BaseSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxOperator;
+import reactor.core.publisher.Mono;
+import reactor.util.context.Context;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Flux operator that auto-completes or auto-abandons messages when control is returned successfully.
+ */
+class FluxAutoComplete<T extends ServiceBusReceivedMessageContext> extends FluxOperator<T, T> {
+    private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete;
+    private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon;
+    private final ClientLogger logger = new ClientLogger(FluxAutoComplete.class);
+
+    FluxAutoComplete(Flux<? extends T> upstream, Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete,
+        Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon) {
+        super(upstream);
+        this.onComplete = Objects.requireNonNull(onComplete, "'onComplete' cannot be null.");
+        this.onAbandon = Objects.requireNonNull(onAbandon, "'onAbandon' cannot be null.");
+    }
+
+    /**
+     * Invoked when a downstream subscriber is interested in objects published from this operator.
+     *
+     * @param coreSubscriber The subscriber interested in the published items.
+     */
+    @Override
+    public void subscribe(CoreSubscriber<? super T> coreSubscriber) {
+        Objects.requireNonNull(coreSubscriber, "'coreSubscriber' cannot be null.");
+
+        final AutoCompleteSubscriber<T> subscriber =
+            new AutoCompleteSubscriber<T>(coreSubscriber, onComplete, onAbandon, logger);
+
+        source.subscribe(subscriber);
+    }
+
+    static final class AutoCompleteSubscriber<T extends ServiceBusReceivedMessageContext> extends BaseSubscriber<T> {
+        private final CoreSubscriber<? super T> downstream;
+        private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete;
+        private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon;
+        private final ClientLogger logger;
+
+        AutoCompleteSubscriber(CoreSubscriber<? super T> downstream,
+            Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete,
+            Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon, ClientLogger logger) {
+            this.downstream = downstream;
+            this.onComplete = onComplete;
+            this.onAbandon = onAbandon;
+            this.logger = logger;
+        }
+
+        @Override
+        protected void hookOnSubscribe(Subscription subscription) {
+            logger.info("Subscription received. Subscribing downstream. {}", subscription);
+            downstream.onSubscribe(this);
+            requestUnbounded();
```

Comment:
Why do we call  `requestUnbounded()`? The downstream should control the flow.

Reply:
There's an exception that gets thrown because this operator itself should request something from upstream. When we get subscribed, we'll pass another subscription that is called and there it can restrict it if it wants.
```
downstreamSubscriber.onError(
    java.lang.IllegalStateException: Can't deliver value due to lack of requests
);
-> at com.azure.messaging.servicebus/com.azure.messaging.servicebus.FluxAutoComplete$AutoCompleteSubscriber.hookOnError(FluxAutoComplete.java:96)
```